source/util/huffman_codec.h

```diff
@@ -0,0 +1,258 @@
+    queue.emplace(root_, "");
+
+    while (!queue.empty()) {
+      const Node* node = queue.front().first;
+      const std::string code = queue.front().second;
+      queue.pop();
+      if (!node->right && !node->left) {
+        out << node->val;
+        if (print_weights)
+            out << " " << node->weight;
+        out << " " << code << std::endl;
+      } else {
+        if (node->left)
+          queue.emplace(node->left, code + "0");
+
+        if (node->right)
+          queue.emplace(node->right, code + "1");
+      }
+    }
+  }
+
+  const std::unordered_map<Val, std::pair<uint64_t, size_t>>&
+      GetEncodingTable() const {
+    return encoding_table_;
+  }
+
+  // Encodes |val| and stores the value in the lower |num_bits| of |bits|.
+  bool Encode(const Val& val, uint64_t* bits, size_t* num_bits) {
+    auto it = encoding_table_.find(val);
+    if (it == encoding_table_.end())
+      return false;
+    *bits = it->second.first;
+    *num_bits = it->second.second;
+    return true;
+  }
+
+  // Reads bits one-by-one using callback |read_bit| until a match is found.
+  // Matching value is stored in |val|. Returns false if the stream terminates
+  // prematurely.
+  bool DecodeFromStream(const std::function<bool(bool*)>& read_bit, Val* val) {
+    Node* node = root_;
+    while (true) {
+      assert(node);
+
+      if (node->left == nullptr && node->right == nullptr) {
+        *val = node->val;
+        return true;
+      }
+
+      bool go_right;
+      if (!read_bit(&go_right))
+        return false;
+
+      if (go_right)
+        node = node->right;
+      else
+        node = node->left;
+    }
+
+    assert (0);
+    return false;
+  }
+
+ private:
+  // Huffman tree node.
+  struct Node {
+    Val val = Val();
+    uint32_t weight = 0;
+    uint32_t id = 0;
+    Node* left = nullptr;
+    Node* right = nullptr;
+  };
+
+  static bool LeftIsBigger(const Node* left, const Node* right) {
+    if (left->weight == right->weight) {
+      assert (left->id != right->id);
+      return left->id > right->id;
+    }
+    return left->weight > right->weight;
+  }
+
+  // Prints subtree (helper function used by PrintTree).
+  static void PrintTreeInternal(std::ostream& out, Node* node, size_t depth) {
+    if (!node)
+      return;
+
+    if (!node->right && !node->left) {
+      out << node->val << std::endl;
+    } else {
+      if (node->right) {
+        out << std::setfill('-') << std::left << std::setw(7) << node->right->weight;
+        PrintTreeInternal(out, node->right, depth + 1);
+      }
+
+      if (node->left) {
+        out << std::string(depth * 7, ' ');
+        out << std::setfill('-') << std::left << std::setw(7) << node->left->weight;
+        PrintTreeInternal(out, node->left, depth + 1);
+      }
+    }
+  }
+
+  // Traverses the Huffman tree and saves paths to the leaves as bit
+  // sequences to encoding_table_.
+  void CreateEncodingTable() {
+    std::queue<std::tuple<Node*, uint64_t, size_t>> queue;
```

Comment:
If all the weights are 0 then everything is ordered by 0 rather than by weights.  Then I think the queue always picks off the nodes with largest IDs to combine them.  So you get a ladder.  ??

This problem goes away if you have as a precondition to the entire table that the weights are assumed to be strictly positive.

Also, you must put the analysis in the *comments* rather than in the review.